File: mcp_gateway/heartbeat.py
#!/usr/bin/env python3
"""
heartbeat — Background heartbeat thread for the gateway.

Sends periodic POST /heartbeat via ModConnection to the game mod.
On first heartbeat, triggers the callback (decompile, etc).
After consecutive failures, marks the mod as disconnected.
"""

import sys
import threading
import time
from typing import Callable

from mcp_gateway.mod_client import ModConnection

Callback = Callable[[dict], None]


class HeartbeatManager:
    DEFAULT_INTERVAL = 5.0
    DEFAULT_MAX_FAILURES = 3

    def __init__(
        self,
        mod_conn: ModConnection,
        workspace_dir: str,
        on_first_heartbeat: Callback | None = None,
        on_heartbeat: Callback | None = None,
        interval: float | None = None,
        max_failures: int | None = None,
    ):
        self.mod: ModConnection = mod_conn
        self.workspace_dir = workspace_dir
        self.on_first_heartbeat = on_first_heartbeat
        self.on_heartbeat = on_heartbeat
        self.interval = max(interval or self.DEFAULT_INTERVAL, 0.1)
        self.max_failures = max(max_failures or self.DEFAULT_MAX_FAILURES, 1)

        self._thread: threading.Thread | None = None
        self._stop = threading.Event()
        self._lock = threading.Lock()

        self._connected = False
        self._first_heartbeat_done = False
        self._consecutive_failures = 0
        self._last_response: dict | None = None
        self._session_id: str | None = None

    @property
    def connected(self) -> bool:
        with self._lock:
            return self._connected

    @property
    def first_heartbeat_done(self) -> bool:
        with self._lock:
            return self._first_heartbeat_done

    @property
    def session_id(self) -> str | None:
        with self._lock:
            return self._session_id

    def start(self):
        if self._thread and self._thread.is_alive():
            return
        self._stop.clear()
        self._thread = threading.Thread(target=self._run, daemon=True, name="heartbeat")
        self._thread.start()

    def stop(self):
        self._stop.set()
        if self._thread:
            self._thread.join(timeout=3)

    def _run(self):
        while not self._stop.is_set():
            ok, resp = self.mod.send_heartbeat(self.workspace_dir)

            with self._lock:
                if ok:
                    self._connected = True
                    self._consecutive_failures = 0
                    self._last_response = resp

                    if not self._first_heartbeat_done:
                        self._first_heartbeat_done = True
                        self._session_id = resp.get("sessionId") if resp else None
                        triggered = True
                    else:
                        triggered = False
                else:
                    self._consecutive_failures += 1
                    if self._consecutive_failures >= self.max_failures:
                        self._connected = False
                    triggered = False

            if resp:
                cb = self.on_first_heartbeat if triggered else self.on_heartbeat
                if cb:
                    try:
                        cb(resp)
                    except Exception as e:
                        print(
                            f"[heartbeat] callback error: {e}",
                            file=sys.stderr,
                            flush=True,
                        )

            self._stop.wait(self.interval)

File: gateway/src/heartbeat.ts
/**
 * heartbeat.ts — background heartbeat for the gateway.
 *
 * Port of `mcp_gateway/heartbeat.py`. Node is single-threaded, so instead of a
 * dedicated daemon thread we use a `setInterval`. On the first successful
 * heartbeat a one-time callback fires; afterwards `onHeartbeat` fires only on
 * tool/reload count changes. After `maxFailures` consecutive failures the mod
 * is marked disconnected.
 *
 * Thread-safety (of the async registry mutations the Python version had to
 * schedule) is handled by the caller: all callbacks are invoked synchronously
 * from this interval tick, and dynamic-tool registration is expected to run on
 * the same synchronous flow.
 */
export interface HeartbeatResponse {
  sessionId?: string;
  toolCount?: number;
  reloadCount?: number;
  [key: string]: unknown;
}

export type HeartbeatCallback = (resp: HeartbeatResponse) => void | Promise<void>;

export interface HeartbeatManagerOptions {
  sendHeartbeat: (workspaceDir: string) => Promise<[boolean, Record<string, unknown>]>;
  workspaceDir: string;
  onFirstHeartbeat?: HeartbeatCallback;
  onHeartbeat?: HeartbeatCallback;
  interval?: number;
  maxFailures?: number;
}

export class HeartbeatManager {
  readonly DEFAULT_INTERVAL = 5;
  readonly DEFAULT_MAX_FAILURES = 3;

  private readonly sendHeartbeat: HeartbeatManagerOptions["sendHeartbeat"];
  private readonly workspaceDir: string;
  private readonly onFirstHeartbeat?: HeartbeatCallback;
  private readonly onHeartbeat?: HeartbeatCallback;
  private readonly intervalMs: number;
  private readonly maxFailures: number;

  private _timer: ReturnType<typeof setInterval> | null = null;
  private _connected = false;
  private _firstHeartbeatDone = false;
  private _consecutiveFailures = 0;
  private _sessionId: string | undefined;

  constructor(opts: HeartbeatManagerOptions) {
    this.sendHeartbeat = opts.sendHeartbeat;
    this.workspaceDir = opts.workspaceDir;
    this.onFirstHeartbeat = opts.onFirstHeartbeat;
    this.onHeartbeat = opts.onHeartbeat;
    this.intervalMs = Math.max(opts.interval ?? this.DEFAULT_INTERVAL, 0.1) * 1000;
    this.maxFailures = Math.max(opts.maxFailures ?? this.DEFAULT_MAX_FAILURES, 1);
  }

  get connected(): boolean {
    return this._connected;
  }

  get firstHeartbeatDone(): boolean {
    return this._firstHeartbeatDone;
  }

  get sessionId(): string | undefined {
    return this._sessionId;
  }

  start(): void {
    if (this._timer !== null) return;
    void this.tick();
    this._timer = setInterval(() => void this.tick(), this.intervalMs);
    // Do not keep the Node process alive solely for the heartbeat.
    if (typeof this._timer.unref === "function") {
      this._timer.unref();
    }
  }

  stop(): void {
    if (this._timer !== null) {
      clearInterval(this._timer);
      this._timer = null;
    }
  }

  private async tick(): Promise<void> {
    const [ok, resp] = await this.sendHeartbeat(this.workspaceDir);
    const hb = (resp ?? {}) as HeartbeatResponse;

    let triggered: "first" | "change" | null = null;
    if (ok) {
      this._connected = true;
      this._consecutiveFailures = 0;
      if (!this._firstHeartbeatDone) {
        this._firstHeartbeatDone = true;
        this._sessionId = hb.sessionId;
        triggered = "first";
      } else {
        triggered = "change";
      }
    } else {
      this._consecutiveFailures += 1;
      if (this._consecutiveFailures >= this.maxFailures) {
        this._connected = false;
      }
    }

    const cb =
      triggered === "first" ? this.onFirstHeartbeat : triggered === "change" ? this.onHeartbeat : null;
    if (cb) {
      try {
        await cb(hb);
      } catch (err) {
        console.error("[heartbeat] callback error:", err);
      }
    }
  }
}
